Failed tests (10815): TestStoreRangeSystemSplits

The following test appears to have failed:

[#10815](https://circleci.com/gh/cockroachdb/cockroach/10815):

```
RdRg@0.000000000,0          10:21:46.969603  184.682µs  sending batch            kv/txn_coord_sender.go:308
RdRg@0.000000000,0          10:21:46.969617  11.254µs   ·meta descriptor lookup  kv/dist_sender.go:562
RdRg@0.000000000,0          10:21:46.969637  144.07µs   ·sending RPC             kv/dist_sender.go:422

I0108 10:21:47.970408 1021 storage/engine/rocksdb.go:137  closing in-memory rocksdb instance
--- FAIL: TestStoreRangeSystemSplits (5.13s)
	testing.go:145: storage/client_split_test.go:664: condition failed to evaluate within 5s: storage/client_split_test.go:658: expected split keys:
		[/Meta2/"\xf9\x03\xe8" /Meta2/"\xf9\x03\xe9" /Meta2/"\xf9\x03\xea" /Meta2/"\xf9\x03\xeb" /Meta2/"\xf9\x03\xec" /Meta2/"\xf9\x03\xed" /Meta2/"\xf9\x03\xee" /Meta2/"\xf9\x03\xef" /Meta2/"\xf9\x03\xf0" /Meta2/"\xf9\x03\xf1" /Meta2/"\xf9\x03\xf2" /Meta2/"\xf9\x03\xf3" /Meta2/"\xf9\x03\xf4" /Meta2/Max]
		but found:
		[/Meta2/Max]
=== RUN   TestSplitSnapshotRace_SplitWins
E0108 10:21:47.971681 1021 gossip/gossip.go:185  different node IDs were set for the same gossip instance (2147483647, 1)
I0108 10:21:47.971843 1021 raft/raft.go:446  [group 1] 1 became follower at term 5
I0108 10:21:47.971891 1021 raft/raft.go:234  [group 1] newRaft 1 [peers: [1], term: 5, commit: 10, applied: 10, lastindex: 10, lastterm: 5]
I0108 10:21:47.971910 1021 raft/raft.go:526  [group 1] 1 is starting a new election at term 5
I0108 10:21:47.971930 1021 raft/raft.go:459  [group 1] 1 became candidate at term 6
--
I0108 10:21:49.426695 1021 storage/engine/rocksdb.go:137  closing in-memory rocksdb instance
I0108 10:21:49.426835 1021 storage/engine/rocksdb.go:137  closing in-memory rocksdb instance
--- PASS: TestLeaderAfterSplit (0.22s)
=== RUN   Example_rebalancing
--- PASS: Example_rebalancing (0.12s)
FAIL
FAIL	github.com/cockroachdb/cockroach/storage	17.101s
=== RUN   TestBatchBasics
I0108 10:21:23.801659 994 storage/engine/rocksdb.go:137  closing in-memory rocksdb instance
--- PASS: TestBatchBasics (0.00s)
=== RUN   TestBatchGet
I0108 10:21:23.802740 994 storage/engine/rocksdb.go:137  closing in-memory rocksdb instance
--- PASS: TestBatchGet (0.00s)
=== RUN   TestBatchMerge
I0108 10:21:23.805450 994 storage/engine/rocksdb.go:137  closing in-memory rocksdb instance
--- PASS: TestBatchMerge (0.00s)
=== RUN   TestBatchProto

```
Please assign, take a look and update the issue accordingly.

I looked at this a little bit since it failed a lot this morning, and I think I have a rough idea what's going on.
The test fails at the first call to `verifySplitsAtTablePrefixes`, where it essentially expects a bunch of splits, but only gets one.

In the log below, the following happens:

* initial config gets gossiped
* it leads to a single split, which is processed 
* while that split is running, the test writes its update, and the callback fires MaybeAdd (and hence shouldQueue) with the yet unsplit replica
* the split completes
* the queue processes again, but the replica added was the left hand of the split `[/Min, /Max)` at the time it was added, now `[/Min, /Table/11)`, and none of the many split keys is in that range, so this is a noop
* nothing else happens since no more callbacks are fired.

Seems straightforward enough to fix the test by forcing all replicas into the split queue in a loop, but this also seems to imply that in real life, we may end up not splitting our ranges correctly, so one of the queue-professionals should take a look at this. I'm disabling the test for now.

```
W0112 06:00:23.416738 84658 storage/store.go:697  system gossip update to [190 140 178 110 9 124 76 141 2 20]
W0112 06:00:23.416877 84658 storage/split_queue.go:72  shouldQueue: [/Min,/Max): have keys [/Table/11/0] based on conf [190 140 178 110 9 124 76 141 2 20] (12)
W0112 06:00:23.416911 84658 storage/split_queue.go:67  shouldQueue: true, 1.000000
W0112 06:00:23.417074 84658 storage/split_queue.go:101  process: [/Min,/Max): have keys [/Table/11/0] based on conf [190 140 178 110 9 124 76 141 2 20] (12)
I0112 06:00:23.417151 84658 storage/split_queue.go:103  splitting range=1 [/Min-/Max) at keys [/Table/11/0]
W0112 06:00:23.422619 84658 storage/store.go:697  system gossip update to [163 31 15 186 80 90 191 12 52 120]
W0112 06:00:23.422749 84658 storage/split_queue.go:72  shouldQueue: [/Min,/Max): have keys [/Table/11/0 /Table/50/0 /Table/51/0 /Table/52/0 /Table/53/0 /Table/54/0 /Table/55/0 /Table/56/0 /Table/57/0 /Table/58/0 /Table/59/0 /Table/60/0 /Table/61/0 /Table/62/0] based on conf [163 31 15 186 80 90 191 12 52 120] (18)
W0112 06:00:23.422768 84658 storage/split_queue.go:67  shouldQueue: true, 1.000000
W0112 06:00:23.423994 84658 storage/replica_command.go:1273  split result: <nil>
W0112 06:00:23.424028 84658 storage/split_queue.go:109  process is done
W0112 06:00:23.424859 84658 storage/split_queue.go:101  process: [/Min,/Table/11): have keys [] based on conf [163 31 15 186 80 90 191 12 52 120] (18)
W0112 06:00:23.424879 84658 storage/split_queue.go:126  process is done
--- FAIL: TestStoreRangeSystemSplits (10.14s)
	testing.go:145: storage/client_split_test.go:680: condition failed to evaluate within 10s: storage/client_split_test.go:673: expected split keys:
		[/Meta2/"\x93" /Meta2/"\xba" /Meta2/"\xbb" /Meta2/"\xbc" /Meta2/"\xbd" /Meta2/"\xbe" /Meta2/"\xbf" /Meta2/"\xc0" /Meta2/"\xc1" /Meta2/"\xc2" /Meta2/"\xc3" /Meta2/"\xc4" /Meta2/"\xc5" /Meta2/"\xc6" /Meta2/Max]
		but found:
		[/Meta2/"\x93" /Meta2/Max]

```

On Tue, Jan 12, 2016 at 6:09 AM, Tobias Schottdorf <notifications@github.com
> wrote:

> I looked at this a little bit since it failed a lot this morning, and I
> think I have a rough idea what's going on.
> The test fails at the first call to verifySplitsAtTablePrefixes, where it
> essentially expects a bunch of splits, but only gets one.
>
> In the log below, the following happens:
>
>    - initial config gets gossiped
>    - it leads to a single split, which is processed
>
> Note that splitting on the initial config is new with
https://github.com/cockroachdb/cockroach/pull/3737 which moved the lease
table out of the "SystemConfig" span. While this was the correct thing to
do, it required fixing assumptions in various storage tests. Looks like it
is also revealing underlying bugs in the split queue.

>
>    - while that split is running, the test writes its update, and the
>    callback fires MaybeAdd (and hence shouldQueue) with the yet unsplit replica
>    - the split completes
>    - the queue processes again, but the replica added was the left hand
>    of the split [/Min, /Max) at the time it was added, now [/Min,
>    /Table/11), and none of the many split keys is in that range, so this
>    is a noop
>    - nothing else happens since no more callbacks are fired.
>
> Seems straightforward enough to fix the test by forcing all replicas into
> the split queue in a loop, but this also seems to imply that in real life,
> we may end up not splitting our ranges correctly, so one of the
> queue-professionals should take a look at this. I'm disabling the test for
> now.
>
> W0112 06:00:23.416738 84658 storage/store.go:697  system gossip update to [190 140 178 110 9 124 76 141 2 20]
> W0112 06:00:23.416877 84658 storage/split_queue.go:72  shouldQueue: [/Min,/Max): have keys [/Table/11/0] based on conf [190 140 178 110 9 124 76 141 2 20] (12)
> W0112 06:00:23.416911 84658 storage/split_queue.go:67  shouldQueue: true, 1.000000
> W0112 06:00:23.417074 84658 storage/split_queue.go:101  process: [/Min,/Max): have keys [/Table/11/0] based on conf [190 140 178 110 9 124 76 141 2 20] (12)
> I0112 06:00:23.417151 84658 storage/split_queue.go:103  splitting range=1 [/Min-/Max) at keys [/Table/11/0]
> W0112 06:00:23.422619 84658 storage/store.go:697  system gossip update to [163 31 15 186 80 90 191 12 52 120]
> W0112 06:00:23.422749 84658 storage/split_queue.go:72  shouldQueue: [/Min,/Max): have keys [/Table/11/0 /Table/50/0 /Table/51/0 /Table/52/0 /Table/53/0 /Table/54/0 /Table/55/0 /Table/56/0 /Table/57/0 /Table/58/0 /Table/59/0 /Table/60/0 /Table/61/0 /Table/62/0] based on conf [163 31 15 186 80 90 191 12 52 120] (18)
> W0112 06:00:23.422768 84658 storage/split_queue.go:67  shouldQueue: true, 1.000000
> W0112 06:00:23.423994 84658 storage/replica_command.go:1273  split result: <nil>
> W0112 06:00:23.424028 84658 storage/split_queue.go:109  process is done
> W0112 06:00:23.424859 84658 storage/split_queue.go:101  process: [/Min,/Table/11): have keys [] based on conf [163 31 15 186 80 90 191 12 52 120] (18)
> W0112 06:00:23.424879 84658 storage/split_queue.go:126  process is done
> --- FAIL: TestStoreRangeSystemSplits (10.14s)
>     testing.go:145: storage/client_split_test.go:680: condition failed to evaluate within 10s: storage/client_split_test.go:673: expected split keys:
>         [/Meta2/"\x93" /Meta2/"\xba" /Meta2/"\xbb" /Meta2/"\xbc" /Meta2/"\xbd" /Meta2/"\xbe" /Meta2/"\xbf" /Meta2/"\xc0" /Meta2/"\xc1" /Meta2/"\xc2" /Meta2/"\xc3" /Meta2/"\xc4" /Meta2/"\xc5" /Meta2/"\xc6" /Meta2/Max]
>         but found:
>         [/Meta2/"\x93" /Meta2/Max]
>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/issues/3698#issuecomment-170878147>
> .
>

--- a/storage/client_split_test.go
+++ b/storage/client_split_test.go
@@ -534,12 +534,13 @@ func TestStoreZoneUpdateAndRangeSplit(t *testing.T) {
 	// Wait for the range to be split along table boundaries.
 	originalRange := store.LookupReplica(roachpb.RKey(keys.UserTableDataMin), nil)
 	var rng *storage.Replica
-	if err := util.IsTrueWithin(func() bool {
+	util.SucceedsWithin(t, splitTimeout, func() error {
 		rng = store.LookupReplica(keys.MakeTablePrefix(descID), nil)
-		return rng.RangeID != originalRange.RangeID
-	}, splitTimeout); err != nil {
-		t.Fatalf("failed to notice range max bytes update: %s", err)
-	}
+		if rng.RangeID == originalRange.RangeID {
+			return util.Errorf("expected new range created by split")
+		}
+		return nil
+	})
 
 	// Check range's max bytes settings.
 	if rng.GetMaxBytes() != maxBytes {

--- a/storage/client_split_test.go
+++ b/storage/client_split_test.go
@@ -516,7 +516,6 @@ func fillRange(store *storage.Store, rangeID roachpb.RangeID, prefix roachpb.Key
 // exceeding zone's RangeMaxBytes.
 func TestStoreZoneUpdateAndRangeSplit(t *testing.T) {
 	defer leaktest.AfterTest(t)
-	t.Skip("#3762")
 	store, stopper := createTestStore(t)
 	config.TestingSetupZoneConfigHook(stopper)
 	defer stopper.Stop()
@@ -577,7 +576,6 @@ func TestStoreZoneUpdateAndRangeSplit(t *testing.T) {
 // split.
 func TestStoreRangeSplitWithMaxBytesUpdate(t *testing.T) {
 	defer leaktest.AfterTest(t)
-	t.Skip("#3749")
 	store, stopper := createTestStore(t)
 	config.TestingSetupZoneConfigHook(stopper)
 	defer stopper.Stop()
@@ -612,7 +610,6 @@ func TestStoreRangeSplitWithMaxBytesUpdate(t *testing.T) {
 // the SystemConfig span.
 func TestStoreRangeSystemSplits(t *testing.T) {
 	defer leaktest.AfterTest(t)
-	t.Skip("#3698")
 	store, stopper := createTestStore(t)
 	defer stopper.Stop()
 

--- a/storage/client_test.go
+++ b/storage/client_test.go
@@ -94,7 +94,7 @@ func createTestStoreWithEngine(t *testing.T, eng engine.Engine, clock *hlc.Clock
 	sCtx.Gossip = gossip.New(rpcContext, gossip.TestBootstrap)
 	sCtx.Gossip.SetNodeID(nodeDesc.NodeID)
 	sCtx.Tracer = tracer.NewTracer(nil, "testing")
-	localSender := storage.NewStores(clock)
+	stores := storage.NewStores(clock)
 	rpcSend := func(_ rpc.Options, _ string, _ []net.Addr,
 		getArgs func(addr net.Addr) proto.Message, _ func() proto.Message,
 		_ *rpc.Context) ([]proto.Message, error) {
@@ -102,7 +102,7 @@ func createTestStoreWithEngine(t *testing.T, eng engine.Engine, clock *hlc.Clock
 		trace := sCtx.Tracer.NewTrace("store", ba)
 		defer trace.Finalize()
 		ctx := tracer.ToCtx(context.Background(), trace)
-		br, pErr := localSender.Send(ctx, *ba)
+		br, pErr := stores.Send(ctx, *ba)
 		if br == nil {
 			br = &roachpb.BatchResponse{}
 		}
@@ -115,8 +115,8 @@ func createTestStoreWithEngine(t *testing.T, eng engine.Engine, clock *hlc.Clock
 	}
 	distSender := kv.NewDistSender(&kv.DistSenderContext{
 		Clock:             clock,
-		RPCSend:           rpcSend,     // defined above
-		RangeDescriptorDB: localSender, // for descriptor lookup
+		RPCSend:           rpcSend, // defined above
+		RangeDescriptorDB: stores,  // for descriptor lookup
 	}, sCtx.Gossip)
 
 	sender := kv.NewTxnCoordSender(distSender, clock, false, nil, stopper)
@@ -130,7 +130,7 @@ func createTestStoreWithEngine(t *testing.T, eng engine.Engine, clock *hlc.Clock
 			t.Fatal(err)
 		}
 	}
-	localSender.AddStore(store)
+	stores.AddStore(store)
 	if bootstrap {
 		if err := store.BootstrapRange(sql.MakeMetadataSchema().GetInitialValues()); err != nil {
 			t.Fatal(err)
@@ -333,7 +333,7 @@ func (m *multiTestContext) rpcSend(_ rpc.Options, _ string, addrs []net.Addr,
 		case *roachpb.RangeKeyMismatchError:
 		case *roachpb.NotLeaderError:
 			if tErr.Leader == nil {
-				// localSender has the range, is *not* the Leader, but the
+				// stores has the range, is *not* the Leader, but the
 				// Leader is not known; this can happen if the leader is removed
 				// from the group. Move the manual clock forward in an attempt to
 				// expire the lease.

--- a/storage/client_test.go
+++ b/storage/client_test.go
@@ -93,6 +93,7 @@ func createTestStoreWithEngine(t *testing.T, eng engine.Engine, clock *hlc.Clock
 	nodeDesc := &roachpb.NodeDescriptor{NodeID: 1}
 	sCtx.Gossip = gossip.New(rpcContext, gossip.TestBootstrap)
 	sCtx.Gossip.SetNodeID(nodeDesc.NodeID)
+	sCtx.ScanMaxIdleTime = splitTimeout / 10
 	sCtx.Tracer = tracer.NewTracer(nil, "testing")
 	stores := storage.NewStores(clock)
 	rpcSend := func(_ rpc.Options, _ string, _ []net.Addr,
@@ -122,6 +123,7 @@ func createTestStoreWithEngine(t *testing.T, eng engine.Engine, clock *hlc.Clock
 	sender := kv.NewTxnCoordSender(distSender, clock, false, nil, stopper)
 	sCtx.Clock = clock
 	sCtx.DB = client.NewDB(sender)
+	sCtx.StorePool = storage.NewStorePool(sCtx.Gossip, clock, storage.TestTimeUntilStoreDeadOff, stopper)
 	sCtx.Transport = storage.NewLocalRPCTransport(stopper)
 	// TODO(bdarnell): arrange to have the transport closed.
 	store := storage.NewStore(*sCtx, eng, nodeDesc)

--- a/storage/queue.go
+++ b/storage/queue.go
@@ -399,8 +399,7 @@ func (bq *baseQueue) processReplica(repl *Replica, clock *hlc.Clock) {
 	if err := bq.impl.process(clock.Now(), repl, cfg); err != nil {
 		bq.eventLog.Errorf("%s: error: %v", repl, err)
 	} else {
-		bq.eventLog.Infof(log.V(2), "%s: done: %0.2fms", repl,
-			time.Since(start).Seconds()*1000)
+		bq.eventLog.Infof(log.V(2), "%s: done: %s", repl, time.Since(start))
 	}
 }
 

